_posts/2025-09-01-semantic-router.md

```diff
@@ -0,0 +1,112 @@
+* **DeepSeek v3.1**: Implements a "single-model dual-mode" design, merging dialogue and reasoning.
+
+In short: the industry is entering an era where **no token should be wasted**.
+
+---
+
+## Recent Research: vLLM Semantic Router
+
+Amid this shift toward hybrid reasoning, we focus on the **open-source inference engine vLLM**.
+
+While vLLM has become the de facto standard for deploying large models, it lacks fine-grained, semantic-level control — the ability to make routing decisions based on meaning rather than query type alone. Developers are often forced to either enable full inference (wasting computation) or disable it entirely (sacrificing accuracy).
+
+To address this, we propose the **vLLM Semantic Router**, which brings GPT-5-style "smart routing" to the open-source ecosystem.
+
+![](/assets/figures/semantic-router/architecture.png)
+
+### Architecture Design
+
+1. **Semantic Classification**: Uses a **ModernBERT** fine-tuned intent classifier to determine whether a query requires inference.
+2. **Smart Routing**:
+
+   * Simple queries → Fast inference mode.
+   * Complex queries → Chain-of-Thought for accurate reasoning.
+3. **High-Performance Engine**: Built with Rust and the Hugging Face Candle framework, enabling high concurrency and zero-copy efficiency.
+4. **Cloud-Native Integration**: Seamlessly integrates with Kubernetes and API Gateways via the Envoy `ext_proc` plugin for enterprise deployments.
+
+Experimental results show:
+
+* **Accuracy**: +10.2%
+* **Latency**: –47.1%
+* **Token Consumption**: –48.5%
+
+In knowledge-intensive areas such as business and economics, accuracy improvements can exceed **20%**.
+
+---
+
+## Project Background
+
+The Semantic Router is not the isolated result of a single paper but a collaborative outcome of sustained community contributions:
+
+* Originally proposed by **Dr. Chen Huamin**, Distinguished Engineer at **Red Hat**, in early **2025** across multiple open-source communities.
+* Iterated and further developed by **Xunzhuo Liu** at **Tencent**, later contributed to the vLLM community.
+* **Dr. Wang Chen** from **IBM Research** and **Dr. Chen Huamin** will present the project at **KubeCon North America 2025**.
+
+The mission is clear: to serve as an **inference accelerator** for open-source large models:
+
+* Preserve accuracy while minimizing unnecessary token usage.
+* Enable seamless switching between "fast" and "slow" thinking modes without fully enabling or disabling inference.
+* Deliver production-ready enterprise integration through native Kubernetes and Envoy support.
+
+The vLLM Semantic Router is therefore not just a research milestone but an **essential bridge for open-source AI infrastructure**, translating **academic innovation into industrial application**.
+
+You can start exploring the project here: [https://github.com/vllm-project/semantic-router](https://github.com/vllm-project/semantic-router).
```

Comment:
can you also add the v0.1 roadmap here and call for contribution?

Reply:
Yes, sounds good, working on it

Reply:
Done